test/unit/transport/GrpcTransport.spec.js

```diff
@@ -0,0 +1,342 @@
+    url = grpcTransport.makeGrpcUrlFromAddress(dapiAddress);
+  });
+
+  describe('#request', () => {
+    let method;
+    let clientClassMock;
+    let requestMessage;
+    let options;
+    let data;
+    let requestFunc;
+
+    beforeEach(function beforeEach() {
+      data = 'result';
+      method = 'method';
+      requestMessage = 'requestMessage';
+      options = {
+        option: 'value',
+      };
+
+      requestFunc = this.sinon.stub().resolves(data);
+
+      clientClassMock = this.sinon.stub().returns({
+        [method]: requestFunc,
+      });
+
+      addressProviderMock.hasLiveAddresses.resolves(true);
+
+      globalOptions = {
+        retries: 1,
+      };
+
+      grpcTransport = new GrpcTransport(
+        createAddressProviderFromOptionsMock,
+        addressProviderMock,
+        globalOptions,
+      );
+    });
+
+    it('should make a request', async () => {
+      const receivedData = await grpcTransport.request(
+        clientClassMock,
+        method,
+        requestMessage,
+        options,
+      );
+
+      expect(receivedData).to.equal(data);
+      expect(createAddressProviderFromOptionsMock).to.be.calledOnceWithExactly(options);
+      expect(clientClassMock).to.be.calledOnceWithExactly(url);
+      expect(requestFunc).to.be.calledOnceWithExactly(requestMessage);
+      expect(grpcTransport.lastUsedAddress).to.deep.equal(dapiAddress);
+    });
+
+    it('should throw unknown error', async () => {
+      const error = new Error('Unknown error');
+
+      requestFunc.throws(error);
+
+      try {
+        await grpcTransport.request(
+          clientClassMock,
+          method,
+          requestMessage,
+          options,
+        );
+
+        expect.fail('should throw error');
+      } catch (e) {
+        expect(e).to.deep.equal(error);
+        expect(createAddressProviderFromOptionsMock).to.be.calledOnceWithExactly(options);
+        expect(clientClassMock).to.be.calledOnceWithExactly(url);
+        expect(requestFunc).to.be.calledOnceWithExactly(requestMessage);
+      }
+    });
+
+    it('should throw MaxRetriesReachedError', async () => {
+      const error = new Error('Internal error');
+      error.code = GrpcErrorCodes.DEADLINE_EXCEEDED;
+
+      requestFunc.throws(error);
+
+      options.retries = 0;
+      try {
+        await grpcTransport.request(
+          clientClassMock,
+          method,
+          requestMessage,
+          options,
+        );
+
+        expect.fail('should throw MaxRetriesReachedError');
+      } catch (e) {
+        expect(e).to.be.an.instanceof(MaxRetriesReachedError);
+        expect(e.getError()).to.equal(error);
+        expect(createAddressProviderFromOptionsMock).to.be.calledOnceWithExactly(options);
+        expect(clientClassMock).to.be.calledOnceWithExactly(url);
+        expect(requestFunc).to.be.calledOnceWithExactly(requestMessage);
+      }
+    });
+
+    it('should throw NoAvailableAddressesForRetry error', async () => {
```

Comment:
should throw NoAvailableAddressesForRetry error if there are no more available addresses to request